lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp

```diff
@@ -0,0 +1,256 @@
+      mapping.map(std::get<0>(operand), std::get<1>(operand));
+    }
+
+    mlir::Operation *clonedOp = builder.clone(*opToHoist, mapping);
+
+    // Add a return operation to the function
+    builder.create<mlir::func::ReturnOp>(opToHoist->getLoc(),
+                                         clonedOp->getResults());
+
+    // Declare the function prototype in the source module
+    localFunc =
+        func::FuncOp::create(opToHoist->getLoc(), functionName, funcType);
+    localFunc.setPrivate(); // Mark as external (no body)
+    sourceModule.push_back(localFunc);
+
+    hoistedFunc->setAttr("arg_ranks", builder.getI64ArrayAttr(ranks));
+  }
+
+  // Replace the original operation with a call to the hoisted function
+  mlir::OpBuilder opBuilder(opToHoist);
+  auto callOp = opBuilder.create<mlir::func::CallOp>(
+      opToHoist->getLoc(), localFunc, opToHoist->getOperands());
+
+  // Replace all results of the original operation with the call results
+  for (auto result : llvm::zip(opToHoist->getResults(), callOp.getResults())) {
+    std::get<0>(result).replaceAllUsesWith(std::get<1>(result));
+  }
+
+  // Erase the original operation
+  opToHoist->erase();
+}
+
+// an analysis class which relies on manual input locations to mark ops to be
+// hoisted, useful for testing
+class TTIRHoistAnalyzeManual {
+public:
+  using HoistOpSet = llvm::SmallVector<llvm::SmallSet<mlir::Operation *, 4>>;
+
+  TTIRHoistAnalyzeManual(mlir::ModuleOp moduleOp,
+                         const llvm::SmallSet<llvm::StringRef, 4> &targetLocs) {
+
+    moduleOp.walk([&](mlir::Operation *nestedOp) {
+      if (const auto nameLoc = dyn_cast<NameLoc>(nestedOp->getLoc())) {
+        if (targetLocs.contains(nameLoc.getName().getValue())) {
+          llvm::SmallSet<mlir::Operation *, 4> opSet;
+          opSet.insert(nestedOp);
+          hoistedOps.push_back(opSet);
+        }
+      }
+    });
+  }
+
+  HoistOpSet getResults() { return hoistedOps; }
+
+private:
+  HoistOpSet hoistedOps;
+};
+
+// Transform pass to hoist specific ops (based on configured analysis pass) into
+// a cpu submodule for later independent lowering
+class TTIRHoistTransform
+    : public impl::TTIRHoistTransformBase<TTIRHoistTransform> {
+public:
+  using impl::TTIRHoistTransformBase<
+      TTIRHoistTransform>::TTIRHoistTransformBase;
+
+  void runOnOperation() final {
+    mlir::ModuleOp moduleOp = getOperation();
+
+    IRRewriter rewriter(&getContext());
+
+    auto loc = moduleOp->getLoc();
+
+    // TODO (#1629): add logic for automatic vs manual flag here when we
+    // have automatic analysis pass
+
+    // only create cpu-module etc. if we actually have ops to hoist
+    if (!hoistLocs.hasValue()) {
+      return;
+    }
+    const auto targetLocs = getLocations();
+    if (targetLocs.empty()) {
+      return;
+    }
+
+    TTIRHoistAnalyzeManual analysisPass(moduleOp, targetLocs);
+    const auto &hoistOpSets = analysisPass.getResults();
```

Comment:
define global alias using HoistOpSet = llvm::SmallVector<llvm::SmallSet<mlir::Operation *, 4>>; and then use it instead of auto

Reply:
I prefer to access via TTIRHoistAnalyze::HoistOpSet because I don't like globals :) but I will remove auto

Reply:
Yea thats fine